### PR TITLE
multi-selection of auras with shift modifier

### PR DESCRIPTION
**Pull request by mrbuds**
It works only if the first aura selected and the second are both in the same group or are top level.

implement #810

### CODE REVIEW

**Comment by emptyrivers** [comment]
This seems to have some scaling issues. Performance drops off quickly, and when selecting a large number of auras (the primary use case i think), the UI can start to glitch out and not render DisplayButtons that it should be rendering.

**Comment by mrbuds** [comment]
Indeed i confirm the problem.

It needs a batch version of frame.PickDisplayMultiple in OptionsFrame.lua

WeakAuras.PickDisplayMultiple() does for each aura:
WeakAuras.EnsureOptions(id);
WeakAuras.ReloadTriggerOptions(tempGroup);
self:FillOptions(displayOptions[tempGroup.id]);

**Comment by mrbuds** [comment]
i'm dumb i did only in 1 part of the code...

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 1]
```diff
@@ -4072,6 +4072,84 @@ function WeakAuras.PickDisplayMultiple(id)
...
+      first = frame.pickedDisplay;
+    end
+    if (first and first ~= target) then
+      -- check that id is in same group as first
+      local firstData = WeakAuras.GetData(first);
+      -- in a group
+      if (firstData.parent) then
+        local sameGroup = false;
+        local group = WeakAuras.GetData(firstData.parent);
+        for index, child in pairs(group.controlledChildren) do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: You can use GetData(target) here and check .parent of that.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 2]
```diff
@@ -4072,6 +4072,84 @@ function WeakAuras.PickDisplayMultiple(id)
...
+          if #batchSelection > 0 then
+            frame:PickDisplayBatch(batchSelection);
+          end
+        end
+      else
+        -- 1st aura selected is top-level, check if target is also top-level and is not a group
+        local targetData = WeakAuras.GetData(target);
+        if (targetData and not targetData.controlledChildren and not targetData.parent) then
+          local batchSelection = {};
+          for index,button in pairs(frame.buttonsScroll.children) do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: I think this must be ipairs

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 3]
```diff
@@ -4072,6 +4072,84 @@ function WeakAuras.PickDisplayMultiple(id)
...
+        local group = WeakAuras.GetData(firstData.parent);
+        for index, child in pairs(group.controlledChildren) do
+          if (child == target) then
+            sameGroup = true;
+            break;
+          end
+        end
+        if (sameGroup) then
+          local batchSelection = {};
+          for index, child in pairs(group.controlledChildren) do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: This must be ipairs.

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 2 above)
Comment (InfusOnWoW) [reply WeakAurasOptions/WeakAurasOptions.lua]: Or well, it doesn't need to be, but it looks better if it is.

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 4]
```diff
@@ -4072,6 +4072,84 @@ function WeakAuras.PickDisplayMultiple(id)
...
+            -- 1st button
+            if (child == target or child == first) then
+              table.insert(batchSelection, child);
+              for i=index+1,#group.controlledChildren do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: Missing spaces around = and after ,

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 2 above)
Comment (InfusOnWoW) [reply WeakAurasOptions/WeakAurasOptions.lua]: Also missing space after ,

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 5]
```diff
@@ -4072,6 +4072,84 @@ function WeakAuras.PickDisplayMultiple(id)
...
+        -- 1st aura selected is top-level, check if target is also top-level and is not a group
+        local targetData = WeakAuras.GetData(target);
+        if (targetData and not targetData.controlledChildren and not targetData.parent) then
+          local batchSelection = {};
+          for index,button in pairs(frame.buttonsScroll.children) do
+            local data = button.data;
+            -- 1st button
+            if (data and (data.id == target or data.id == first)) then
+              table.insert(batchSelection, data.id);
+              for i=index+1,#frame.buttonsScroll.children do
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: Missing spaces

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 1 above)
Comment (mrbuds) [reply WeakAurasOptions/WeakAurasOptions.lua]: I could do 
      local firstData = WeakAuras.GetData(first);
      local targetData = WeakAuras.GetData(target);
and then compare firstData.parent == targetData.parent (which is more readable and concise)
But after checking both items are in the same group, i still need to iterate over the group children, so another GetData() on the parent

The way i did it save a GetData() call, i have not checked the complexity of this call, do you think their is not enough performance gain doing what i did for what we loose in readability ?

Comment (InfusOnWoW) [reply WeakAurasOptions/WeakAurasOptions.lua]: GetData is cheap, iterating over all childrens is not as cheap. So using that to figure out whether they are the same parent is probably better.

**Comment by mrbuds** [comment]
fixed a bug, and "Duplicate All" and re-opening option with a multi-selection should be faster

**Comment by mrbuds** [comment]
shiiiiit since last commit, when the right click on a multi-selection act as only 1 aura is selected

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 6]
```diff
@@ -4072,6 +4065,71 @@ function WeakAuras.PickDisplayMultiple(id)
   frame:PickDisplayMultiple(id);
 end
 
+function WeakAuras.PickDisplayMultipleShift(target)
+  if (frame.pickedDisplay) then
+    -- get first aura selected
+    local first;
+    if (WeakAuras.IsPickedMultiple()) then
+      first = tempGroup.controlledChildren[1];
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: I think using the last one here is actually better. The controlledChildren of tempGroup are in the order in which they were picked, so using the last one uses the last one that was selected, which matches what's normal for shift selection

File: WeakAurasOptions/WeakAurasOptions.lua [hunk 7]
```diff
@@ -4072,6 +4065,71 @@ function WeakAuras.PickDisplayMultiple(id)
...
+                -- last button: stop selection
+                if (current == target or current == first) then
+                  break;
+                end
+              end
+              break;
+            end
+          end
+        else
+          -- top-level
```
Comment (InfusOnWoW) [inline WeakAurasOptions/WeakAurasOptions.lua]: You might want to check for firstData.parent == nil and targetData.parent == nil here, to confirm that we are handling the case of both being top level. 

The code works without that, but the comment is wrong, the else also currently handles the case of mixed top-levelness

File: WeakAurasOptions/WeakAurasOptions.lua (hunk 6 above)
Comment (mrbuds) [reply WeakAurasOptions/WeakAurasOptions.lua]: Indeed, i wasn't thinking enough about how shift-selection works with some auras already selected with ctrl. I also need to change frame.PickDisplayBatch to not do any PickDisplay()

**Comment by mrbuds** [comment]
Thanks Infus it's much better using the last element of previous selection
example
Just one comment before someone report it as a bug, as you can see it in this example, the shift-selection as it is implemented doesn't show previews of auras selected like left click or ctrl-click selection does.

**Comment by Stanzilla** [comment]
```Just one comment before someone report it as a bug, as you can see it in this example, the shift-selection as it is implemented doesn't show previews of auras selected like left click or ctrl-click selection does.
```

I see this as a problem, they all should behave the same

**Comment by mrbuds** [comment]
wait wait wait .. i said this after watching the (cropped) gif without actually checking in game, the preview of all auras is fine

**Pull request merged**